library/core/src/ptr/non_null.rs

```diff
@@ -631,6 +630,9 @@ impl<T: ?Sized> NonNull<T> {
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_allow_const_fn_unstable(unchecked_neg)]
+    #[kani::requires(count.checked_mul(core::mem::size_of::<T>()).is_some())] // Prevent offset overflow
+    #[kani::requires(count * core::mem::size_of::<T>() <= isize::MAX as usize)]
+    #[kani::ensures(|result: &NonNull<T>| result.as_ptr() == self.as_ptr().offset(-(count as isize)))]
```

Comment:
add post-condition that asserts the resulted pointer is pointing to the same object as the original pointer